### PR TITLE
Update accelerate config and launch for mpirun

### PR DESCRIPTION
**Pull request by dmsuehir**
# What does this PR do?

> Note this PR is just for review purposes (after internal review I will file an issue and PR to the huggingface/accelerate repo)

This PR updates accelerate launch to abstract out the call to `mpirun`. Previously, the user would need to call mpirun along with accelerate launch like:
```
export CCL_WORKER_COUNT=1
export MASTER_ADDR=xxx.xxx.xxx.xxx #node0 ip
export CCL_ATL_TRANSPORT=ofi
mpirun -f hostfile -n 16 -ppn 4 accelerate launch examples/nlp_example.py
```
After this update, the same call would look like:
```
accelerate launch examples/nlp_example.py
```

Accelerate config already had a lot of the information needed for `mpirun` like the primary node IP address, the number of machines, and the number of processes. I updated "accelerate config" to include a few more questions that were needed to more the remaining information:
```
Do you want accelerate to launch mpirun? [yes/NO]: yes
Please enter the path to the hostfile to use with mpirun [~/hostfile]: ~/hostfile
Enter the number of oneCCL worker threads [1]: 1
```
If the user answers "no" when asked if they want accelerate to launch mpirun, then the launch will have the same behavior as it previously did.

I also noticed an error when running the `examples/nlp_example.py` with IPEX "The optimizer should not be given for inference mode" (similar to [this issue](https://github.com/intel/neural-compressor/issues/1520)). The fix is to set the model to training model before `accelerate.prepare` (like [this](https://github.com/intel/neural-compressor/pull/1525/files)).

### CODE REVIEW

File: examples/nlp_example.py [hunk 1]
```diff
@@ -150,7 +150,7 @@ def training_function(config, args):
     # Prepare everything
     # There is no specific order to remember, we just need to unpack the objects in the same order we gave them to the
     # prepare method.
-
+    model.train()
```
Comment (okhleif-10) [inline examples/nlp_example.py]: why was this not here before?

Comment (okhleif-10) [reply examples/nlp_example.py]: it looks like they train in a loop in line 160, so wondering why this is needed

Comment (dmsuehir) [reply examples/nlp_example.py]: Oh yeah, good question. It needs to happen before `accelerate.prepare`, otherwise IPEX throws an exception saying that the optimizer should not be passed for inference mode (since the model was not set to training mode yet, it thinks we are trying to do inference). We saw the same thing with TLT notebooks/scripts before. Others may not have hit this error yet, if they haven't been using this example script with IPEX.

Comment (okhleif-10) [reply examples/nlp_example.py]: Got it. You should leave a comment there in your PR when you submit it to HF

File: docs/source/basic_tutorials/troubleshooting.md [hunk 2]
```diff
@@ -127,6 +127,18 @@ if accelerator.check_breakpoint():
 This is a known issue. On Linux with kernel version < 5.5, hanging processes have been reported. To avoid 
 encountering this problem, we recommend upgrading your system to a later kernel version.
 
+### Hanging when running distributed training using MPI
+
+If your distributed CPU training job using MPI is hanging, ensure that you have passwordless SSH setup (using keys)
```
Comment (ashahba) [inline docs/source/basic_tutorials/troubleshooting.md]: passwordless SSH -> 
[passwordless SSH](https://www.open-mpi.org/faq/?category=rsh#ssh-keys)

File: tests/test_configs/mpi_multicpu.yaml [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  mpirun_ccl: '1'
+  mpirun_hostfile: /home/user/hostfile
+num_machines: 4
+num_processes: 16
+rdzv_backend: static
+same_network: true
+tpu_env: []
+tpu_use_cluster: false
+tpu_use_sudo: false
+use_cpu: true
```
Comment (ashahba) [inline tests/test_configs/mpi_multicpu.yaml]: Missing end of line char

**Review by ashahba — APPROVED** [review]
LGTM!

**Comment by sywangyi** [comment]
for ipex optimize issue. a PR is merged, see https://github.com/huggingface/accelerate/commit/9c1d5bac1555ea8ba1f1bffee616cb2114055243
